snapshotter/pkg/dmsetup/dmsetup.go

```diff
@@ -0,0 +1,310 @@
+		return deviceName
+	}
+
+	return DevMapperDir + deviceName
+}
+
+// BlockDeviceSize returns size of block device in bytes
+func BlockDeviceSize(devicePath string) (uint64, error) {
+	data, err := exec.Command("blockdev", "--getsize64", "-q", devicePath).CombinedOutput()
+	output := string(data)
+	if err != nil {
+		return 0, errors.Wrapf(err, output)
+	}
+
+	output = strings.TrimSuffix(output, "\n")
+	return strconv.ParseUint(output, 10, 64)
+}
+
+func dmsetup(args ...string) (string, error) {
+	data, err := exec.Command("dmsetup", args...).CombinedOutput()
+	output := string(data)
+	if err != nil {
+		// Try find Linux error code otherwise return generic error with dmsetup output
+		if errno, ok := tryGetUnixError(output); ok {
+			return "", errno
+		}
+
+		return "", errors.Wrapf(err, "dmsetup %s\nerror: %s\n", strings.Join(args, " "), output)
+	}
+
+	output = strings.TrimSuffix(output, "\n")
+	output = strings.TrimSpace(output)
+
+	return output, nil
+}
+
+// tryGetUnixError tries to find Linux error code from dmsetup output
+func tryGetUnixError(output string) (unix.Errno, bool) {
+	// It's useful to have Linux error codes like EBUSY, EPERM, ..., instead of just text.
+	// Unfortunately there is no better way than extracting/comparing error text.
+	text := parseDmsetupError(output)
+	if text == "" {
+		return 0, false
+	}
+
+	errTableInit.Do(func() {
+		// Precompute map of <text>=<errno> for optimal lookup
+		errTable = make(map[string]unix.Errno)
+		for errno := unix.EPERM; errno <= unix.EHWPOISON; errno++ {
+			errTable[errno.Error()] = errno
+		}
+	})
+
+	err, ok := errTable[text]
+	return err, ok
+}
+
+// dmsetup returns error messages in format:
+// 	device-mapper: message ioctl on <name> failed: File exists\n
+// 	Command failed\n
+// parseDmsetupError extracts text between "failed: " and "\n"
+func parseDmsetupError(output string) string {
+	re := regexp.MustCompilePOSIX("failed: [^:]+$")
+	str := re.FindString(output)
+
+	// Strip "failed: " prefix
+	str = strings.TrimPrefix(str, "failed: ")
+
+	str = strings.ToLower(str)
+	return str
+}
+
+// makeThinPoolMapping makes thin-pool table entry
+func makeThinPoolMapping(dataFile, metaFile string, blockSizeSectors uint32) (string, error) {
+	dataDeviceSizeBytes, err := BlockDeviceSize(dataFile)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to get block device size: %s", dataFile)
+	}
+
+	// Thin-pool mapping target has the following format:
+	// start - starting block in virtual device
+	// length - length of this segment
+	// metadata_dev - the metadata device
+	// data_dev - the data device
+	// data_block_size - the data block size in sectors
+	// low_water_mark - the low water mark, expressed in blocks of size data_block_size
+	// feature_args - the number of feature arguments
+	// args
+	lengthSectors := dataDeviceSizeBytes / SectorSize
+	target := fmt.Sprintf("0 %d thin-pool %s %s %d 32768 1 skip_block_zeroing", lengthSectors, metaFile, dataFile, blockSizeSectors)
```

Comment:
I'd recommend doing a bit more here:

(top of the file, or directly above the function)
```go
const (
	lowWaterMark = 32768 // Picked as low water mark because [reason]
	skipZeroing  = "skip_block_zeroing" // Skipping zeroing to reduce latency for device creation
)
```

```go
fmt.Sprintf("0 %d thin-pool %s %s %d %d 1 %s",
	lengthSectors,
	metaFile,
	dataFile,
	blockSizeSectors
	lowWaterMark,
	skipZeroing)
```

Moving these things to named constants lets you leave comments on why we picked them, to accompany your existing comment about the command structure.